components/camera/videosource/webcam_e2e_test.go

```diff
@@ -0,0 +1,121 @@
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, len(webcams.Webcams), test.ShouldBeZeroValue)
+}
+
+func TestWebcamDiscovery(t *testing.T) {
+	config, err := vcamera.Builder().
+		NewCamera(0, "Lo Res Webcam", vcamera.Resolution{Width: 640, Height: 480}).
+		NewCamera(1, "Hi Res Webcam", vcamera.Resolution{Width: 1280, Height: 720}).
+		Stream()
+
+	test.That(t, err, test.ShouldBeNil)
+	defer config.Shutdown()
+
+	reg, ok := resource.LookupRegistration(camera.API, resource.DefaultModelFamily.WithModel("webcam"))
+	test.That(t, ok, test.ShouldBeTrue)
+
+	ctx := context.Background()
+	logger := golog.NewTestLogger(t)
+	discoveries, err := reg.Discover(ctx, logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	webcams, ok := discoveries.(*pb.Webcams)
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, len(webcams.Webcams), test.ShouldEqual, 2)
+
+	// sort by name
+	sort.Slice(webcams.Webcams, func(i, j int) bool {
+		return -1 == strings.Compare(webcams.Webcams[i].Name, webcams.Webcams[j].Name) //nolint:stylecheck
+	})
+
+	webcam := webcams.Webcams[0]
+	test.That(t, webcam.Name, test.ShouldEqual, "Hi Res Webcam")
+	test.That(t, webcam.Properties[0].WidthPx, test.ShouldEqual, 1280)
+	test.That(t, webcam.Properties[0].HeightPx, test.ShouldEqual, 720)
+
+	webcam = webcams.Webcams[1]
+	test.That(t, webcam.Name, test.ShouldEqual, "Lo Res Webcam")
+	test.That(t, webcam.Properties[0].WidthPx, test.ShouldEqual, 640)
+	test.That(t, webcam.Properties[0].HeightPx, test.ShouldEqual, 480)
+}
+
+func newWebcamConfig(name, path string) resource.Config {
+	conf := resource.NewEmptyConfig(
+		resource.NewName(camera.API, name),
+		resource.DefaultModelFamily.WithModel("webcam"),
+	)
+	conf.ConvertedAttributes = &videosource.WebcamConfig{Path: path}
+	return conf
+}
+
+func TestWebcamGetImage(t *testing.T) {
+	config, err := vcamera.Builder().
+		NewCamera(0, "Lo Res Webcam", vcamera.Resolution{Width: 640, Height: 480}).
+		NewCamera(1, "Hi Res Webcam", vcamera.Resolution{Width: 1280, Height: 720}).
+		Stream()
+
+	test.That(t, err, test.ShouldBeNil)
+	defer config.Shutdown()
+
+	logger := golog.NewTestLogger(t)
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	conf := newWebcamConfig("cam1", "video0")
+	cam1, err := videosource.NewWebcam(cancelCtx, nil, conf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer cam1.Close(cancelCtx)
+
+	stream, err := cam1.Stream(cancelCtx)
+	test.That(t, err, test.ShouldBeNil)
+
+	img, _, err := stream.Next(cancelCtx)
+	test.That(t, err, test.ShouldBeNil)
+
+	test.That(t, img.Bounds().Dx(), test.ShouldEqual, 640)
+	test.That(t, img.Bounds().Dy(), test.ShouldEqual, 480)
+
+	conf = newWebcamConfig("cam2", "video1")
+	cam2, err := videosource.NewWebcam(cancelCtx, nil, conf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer cam2.Close(cancelCtx)
+
+	stream, err = cam2.Stream(cancelCtx)
+	test.That(t, err, test.ShouldBeNil)
+
+	img, _, err = stream.Next(cancelCtx)
```

Comment:
nit: these next's should probably call the cleanup functions that are being ignored just for completeness

Reply:
Done